modules/kibana/src/main/java/org/elasticsearch/kibana/KibanaPlugin.java

```diff
@@ -0,0 +1,224 @@
+            // apis needed to perform migrations... ideally these will go away
+            new KibanaWrappedRestHandler(new RestCreateIndexAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestGetAliasesAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestIndexPutAliasAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestRefreshAction(), allowedIndexPatterns),
+
+            // apis needed to access saved objects
+            new KibanaWrappedRestHandler(new RestGetAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestMultiGetAction(settings), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestSearchAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestBulkAction(settings), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestDeleteAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestDeleteByQueryAction(), allowedIndexPatterns),
+
+            // api used for testing
+            new KibanaWrappedRestHandler(new RestUpdateSettingsAction(), allowedIndexPatterns),
+
+            // apis used specifically by reporting
+            new KibanaWrappedRestHandler(new RestGetIndicesAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestIndexAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestUpdateAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestSearchScrollAction(), allowedIndexPatterns),
+            new KibanaWrappedRestHandler(new RestClearScrollAction(), allowedIndexPatterns)
+        );
+
+    }
+
+    @Override
+    public List<Setting<?>> getSettings() {
+        return List.of(KIBANA_INDEX_NAMES_SETTING);
+    }
+
+    static class KibanaWrappedRestHandler extends BaseRestHandler.Wrapper {
+
+        private final List<String> allowedIndexPatterns;
+
+        KibanaWrappedRestHandler(BaseRestHandler delegate, List<String> allowedIndexPatterns) {
+            super(delegate);
+            this.allowedIndexPatterns = allowedIndexPatterns;
+        }
+
+        @Override
+        public String getName() {
+            return "kibana_" + super.getName();
+        }
+
+        @Override
+        public List<Route> routes() {
+            return super.routes().stream().map(route -> new Route(route.getMethod(), "/_kibana" + route.getPath()))
+                .collect(Collectors.toUnmodifiableList());
+        }
+
+        @Override
+        protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+            client.threadPool().getThreadContext().allowSystemIndexAccess(allowedIndexPatterns);
+            return super.prepareRequest(request, new IndexLimitingNodeClient(client, allowedIndexPatterns));
+        }
+    }
+
+    static class IndexLimitingNodeClient extends NodeClient {
```

Comment:
I am not ready to break this out just yet because there are no guarantees that its functional beyond the APIs used by this plugin. Also, I still have some reservations about whether I should do it this way or handle it elsewhere.